File: src/py/test00.py
import serial
import struct
import time
import sys
import struct

ser = serial.Serial(
    port='/dev/ttyUSB0',
    baudrate=9600,
    bytesize=serial.EIGHTBITS,
    parity=serial.PARITY_NONE,
#    stopbits=serial.STOPBITS_ONE,
    timeout=0 # Read timeout in seconds
)

try:
    if ser.isOpen():
        print(f"Serial port {ser.port} opened successfully.")

        byte_list0 = [0x0A, 0x03, 0x79, 0x18, 0x00, 0x07, 0x9c, 0x28] # handshake?
        byte_list1 = [0x0A, 0x03, 0x75, 0x30, 0x00, 0x1B, 0x1E, 0xB9] # responce 59 byte
        byte_list2 = [0x0A, 0x03, 0x79, 0x18, 0x00, 0x0A, 0x5D, 0xED] # responce 25 byte

        ser.write(serial.to_bytes(byte_list1))
        print(f"Sent byte list: {byte_list1}")

        time.sleep(0.5)

# Read binary data
        data1 = ser.read(128) # Read up to 59 bytes

        if data1:
            data1_size = sys.getsizeof(data1)
            print(f"Received data: {data1.hex()}") # Print as hex string
            print(f"Data size: {data1_size}") # Print as hex string
            # B - uint8, H - uint16, I - uint32, Q - uint64
            format_string = '>BHHHHHHHHHHHHHHHHHHHHHHHHHHHI'
            unpacked_data1 = struct.unpack(format_string, data1)
            gridVoltage = unpacked_data1[7]*0.1; #V
            gridFreq = unpacked_data1[8]*0.1;    #Hz
            outVoltage = unpacked_data1[9]*0.1;  #V
            outFreq = unpacked_data1[10]*0.1;    #Hz
            print(f"Work State:       {unpacked_data1[4]}")
            print(f"Battery Class:    {unpacked_data1[5]} Volt")
            print(f"Rated Power:      {unpacked_data1[6]} Watt")
#            print(f"Grid Voltage: {gridVoltage:.1f} Volt {hex(unpacked_data1[7])}") #debug example with hex
            print(f"Grid Voltage:     {gridVoltage:.1f} Volt")
            print(f"Grid Frequence:   {gridFreq:.1f} Hertz")
            print(f"Outзut Voltage:   {outVoltage:.1f} Volt")
            print(f"Output Frequency: {outFreq:.1f} Hertz")

        else:
            print("No data received.")

    else:
        print("Could not open serial port.")

except serial.SerialException as e:
    print(f"Error: {e}")

finally:
    if ser.isOpen():
        ser.close()
#        print(f"Serial port {ser.port} closed.")

File: src/c/test00.c
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <termios.h> // For serial port settings
#include <unistd.h>  // For write()

int main() {
    int serial_port = open("/dev/ttyUSB0", O_RDWR | O_NOCTTY | O_NDELAY);

    if (serial_port < 0) {
        perror("Error opening serial port");
        return 1;
    }

    // Configure serial port settings (e.g., baud rate, data bits, parity)
    struct termios tty;
    if (tcgetattr(serial_port, &tty) != 0) {
        perror("Error getting serial port attributes");
        close(serial_port);
        return 1;
    }

    cfsetospeed(&tty, B9600); // Set output baud rate to 9600
    cfsetispeed(&tty, B9600); // Set input baud rate to 9600

    tty.c_cflag &= ~PARENB;   // No parity
    tty.c_cflag &= ~CSTOPB;   // 1 stop bit
    tty.c_cflag &= ~CSIZE;    // Clear data size bits
    tty.c_cflag |= CS8;       // 8 data bits
    tty.c_cflag &= ~CRTSCTS;  // Disable hardware flow control
    tty.c_cflag |= CREAD | CLOCAL; // Enable receiver, ignore modem control lines

    tty.c_lflag &= ~ICANON;   // Disable canonical mode (raw input)
    tty.c_lflag &= ~ECHO;     // Disable echo
    tty.c_lflag &= ~ECHOE;    // Disable erase character echo
    tty.c_lflag &= ~ECHONL;   // Disable newline echo

    tty.c_iflag &= ~(IXON | IXOFF | IXANY); // Disable software flow control
    tty.c_iflag &= ~(IGNBRK | BRKINT | PARMRK | ISTRIP | INLCR | IGNCR | ICRNL); // Disable special handling of received bytes

    tty.c_oflag &= ~OPOST;    // Disable output processing

    tty.c_cc[VMIN] = 0;       // Non-blocking read
    tty.c_cc[VTIME] = 15;      // 1.5 seconds read timeout

    if (tcsetattr(serial_port, TCSANOW, &tty) != 0) {
        perror("Error setting serial port attributes");
        close(serial_port);
        return 1;
    }

    // Binary data to send

    unsigned char data[] = {0x0A, 0x03, 0x75, 0x30, 0x00, 0x1B, 0x1E, 0xB9};
    int data_len = sizeof(data);

    // Write the binary data
    int bytes_written = write(serial_port, data, data_len);

    if (bytes_written < 0) {
        perror("Error writing to serial port");
    } else if (bytes_written != data_len) {
        fprintf(stderr, "Warning: Only %d of %d bytes written.\n", bytes_written, data_len);
    } else {
        printf("Successfully wrote %d bytes of binary data.\n", bytes_written);
    }

//    getchar(); //press Enter
    sleep(1);

    unsigned char read_buf[256];
    int num_bytes = read(serial_port, &read_buf, sizeof(read_buf));

    if (num_bytes < 0) {
        perror("Error reading from serial port");
    } else if (num_bytes > 0) {
        printf("Received %d bytes: ", num_bytes);
        for (int i = 0; i < num_bytes; i++) {
            printf("0x%02X ", read_buf[i]);
        }
        printf("\n");
    } else {
        printf("No data received.\n");
    }
    close(serial_port);
    return 0;
}
